### DISCUSSION TITLE
Respond to user interaction with the canvas

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-canvas-panel#16] [How Might We]
Types of interaction:

 - Movement - panning and zooming
 - Clicking a particular point
 - Clicking a particular point that is the target of an annotation that the component has rendered
 - Box selection
 - Shape selection
 - Text selection (#15)

_Any_ user interaction might be something the developer wants to respond to for some reason
_Many_ user interactions aren't important and will be left unobserved by the developer's code in most scenarios (e.g., panning and zooming actions)
_Some_ click interactions could be left to the component - e.g., #13 a click on a rendered hyperlink could just bubble up to browser and cause a page navigation, _or_ developer might want to handle this explicitly through handling the event
_Some_ user interactions are meaningless unless the component is in a particular mode and being used to accept user input (e.g., in an annotation tool).

(Needs expanding out into different scenarios)

**Comment by stephenwf** [comment]
Going to put a demo together for this in Atlas. But the short answer here is yes. 

Since Atlas might be rendered on a single `<canvas />` element or `<div />` or anything really - we needed a way pass events down through the model from a single surface (like a canvas element) and then based on what's visible, dispatch those to the correct items.

The "Atlas" object model currently supports the following events on individual "objects" painted (images, boxes, etc.).
```js
<worldImage
    // Verified
    onPointerMove={handler('onPointerMove')}
    onMouseMove={handler('onMouseMove')}
    onMouseEnter={handler('onMouseEnter')}
    onPointerEnter={handler('onPointerEnter')}
    onClick={handler('onClick')}
    onDragStart={handler('onDragStart')}
    onMouseDown={handler('onMouseDown')}
    onMouseLeave={handler('onMouseLeave')}
    onMouseUp={handler('onMouseUp')}
    onPointerDown={handler('onPointerDown')}
    onPointerLeave={handler('onPointerLeave')}
    onPointerUp={handler('onPointerUp')}
    onScroll={handler('onScroll')} // @todo possibly impossible.
    onWheel={handler('onWheel')}
    onTouchCancel={handler('onTouchCancel')}
    onTouchEnd={handler('onTouchEnd')}
    onTouchMove={handler('onTouchMove')}
    onTouchStart={handler('onTouchStart')}
    // => Some issues.
    onDrag={handler('onDrag')} // @todo some strange behaviour if the drag goes outside the bounds.
    onDragEnd={handler('onDragEnd')} // @todo does not end drag if not moused over
    onMouseOut={handler('onMouseOut')} // @todo this should be delivered to items CONTAINED
    onPointerOut={handler('onPointerOut')} // @todo this should be delivered to items CONTAINED
    onMouseOver={handler('onMouseOver')} // @todo this has slightly different behaviour (like out)
    onPointerOver={handler('onPointerOver')} // @todo this has slightly different behaviour (like out)
    onPointerCancel={handler('onPointerCancel')} // @todo unable to verify - should be working.
    // => Not yet implemented.
    // onDragEnter={handler('onDragEnter')} // @todo Fired when a dragged element enters a valid drop target.
    // onDragExit={handler('onDragExit')} // @todo Fired when a dragged element leaves a valid drop target.
    // onDragOver={handler('onDragOver')} // @todo Fired when dragging an element over a drop target. (move)

    uri="http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png"
    target={{ width: 1200, height: 1800 }}
    display={{ width: 1200, height: 1800 }}
  />
```

The drag over events are not yet implemented, as those are for editable canvases where you are dragging items in the space over other items. Some work, but do not match the DOM specification exactly. Scroll events do not make sense here as there is no overflow on individual items to scroll (onWheel is distinct). However I think the available events cover the mentioned use-cases. Text selection would be handled natively.

Each event is the proxied native event (e.g. PointerEvent) with an additional property:
```
{
  atlas: { x: number; y: number }
}
```

Which indicates where the event was triggered. Events are pooled for performance, so the following would not work:
```js
let prevPosition;

someBox.addEventListener((e) => {
  if (prevPosition.x !== e.atlas.x) {
     // never hit as e.atlas is reused.
  }

  prevPosition = e.atlas;
});
```

**Reply by tomcrane** [reply]
Discussion... These are Atlas events which are not IIIF-specific. 
CanvasPanel is about events on content linked to the canvas via annotations, using IIIF, so the developer interaction with user events on the Canvas is in IIIF and Anno terms, potentially using the helpers mentioned in https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/33

```html
<canvas-panel id="cp" />

<script>
const cp = document.getElementById("cp");
cp.addEventListener("click", e => {
   // this e has been decorated with extras and strongly typed stuff
   // is `e` a linking anno? What does it link to? Get resolved target, do something
});
</script>
```

**Reply by stephenwf** [reply]
Additionally, does that extend to mousemove or other high-frequency events that would need optimisation. How does the z-index come in for overlapping things. When you hover over a linking annotation inside of an image annotation, inside of the boundary of a canvas - is that 3 events or one. 

Needs some design on what the event should be and how granular they are.

Wide net, with the `e` variable packed with values
```js
cp.addEventListener('click', e => { ... })
```

More granular, with less variance in the `e`:
```js
cp.addEventListener('click-annotation', e => { ... })
```

Very granular:
```js
cp.addEventListener('click-linking-annotation', e => { ... })
```

**Reply by tomcrane** [reply]
These all look like useful tools for the developer

**Reply by tomcrane** [reply]
To clarify... _inside_ canvas panel, it is using Atlas components, `Box` etc, wiring up clicks and other handlers.
But the event interface exposed by CP hides all this and exposes CP events for my code to handle.

Event data might return helper objects for the annotations (#33) for ease of access to them.

If I need to, I can get at the underlying Atlas model from the IIIF (?) E.g., to take over some aspect of styling?

**Comment by tomcrane** [comment]
See events writeup at https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/45#discussioncomment-1164643

Events fired by DisplayAnnotation instances visible on the canvas will be the source of many events the developer will want to handle.
In an annotation creation context, off-canvas form events will interact with the DisplayAnnotation events.
